pkg/encryption/encryption.go

```diff
@@ -0,0 +1,161 @@
+
+package encryption
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/pbkdf2"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/hex"
+	"io"
+	"os"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+const (
+	iteration    int = 10000 // Openssl makes this number of iterations by default when encrypting/decrypting with pbdkf2
+	split        int = 48    // Number of bytes needed to get key and iv from pbkdf2. 32 on key and rest on iv
+	saltSize     int = 8     // Salt size by default in openssl
+	passwordSize int = 16    // Size of password in bytes when generating random password
+)
+
+type Options struct {
+	Encryption bool
+	JustKey    bool
+	Pass       string
+	Filepath   string
+}
+
+func (e *Options) NewWriter(file io.Writer) (*cipher.StreamWriter, error) {
+	var err error
+	salt := make([]byte, saltSize)
+	if _, err := io.ReadFull(rand.Reader, salt); err != nil {
+		return nil, errors.Wrap(err, "Failed to generate salt")
+	}
+	if e.Pass == "" {
+		log.Debug().Msg("Password not provided, generating new")
+		err := e.generatePassword()
+		if err != nil {
+			return nil, errors.Wrap(err, "Failed to generate random password")
+		}
+	}
+
+	pbkdf, err := pbkdf2.Key(sha256.New, e.Pass, salt, iteration, split)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to generate key from pass")
+	}
+	key := pbkdf[:32]
+	iv := pbkdf[32:]
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to generate cipher")
+	}
+
+	stream := cipher.NewCTR(block, iv)
+	writer := &cipher.StreamWriter{S: stream, W: file}
+
+	e.writeSaltToFile(file, salt)
+
+	return writer, nil
+}
+
+func (e *Options) GetReader(r io.Reader) (*cipher.StreamReader, error) {
+	var err error
+	salt := make([]byte, saltSize+8) //nolint:mnd
+	_, err = r.Read(salt)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to get salt")
+	}
+	salt = salt[8:]
+
+	if e.Pass == "" {
+		return nil, errors.New("Password not provided, please provide password")
+	}
+
+	pbkdf, err := pbkdf2.Key(sha256.New, e.Pass, salt, iteration, split)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to generate key from pass")
+	}
+	key := pbkdf[:32]
+	iv := pbkdf[32:]
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to generate cipher")
+	}
+
+	stream := cipher.NewCTR(block, iv)
+	reader := &cipher.StreamReader{S: stream, R: r}
+
+	return reader, nil
+}
+
+func (e *Options) writeSaltToFile(w io.Writer, salt []byte) {
+	prefix := []byte("Salted__")
+	prefix = append(prefix, salt...)
+	_, err := w.Write(prefix)
+	if err != nil {
+		panic("Failed to write salt")
+	}
+}
+
+func (e *Options) OutputPass() error {
+	if !e.Encryption {
+		return nil
+	}
+
+	if e.JustKey {
+		wr := zerolog.ConsoleWriter{
+			Out:     os.Stderr,
+			NoColor: true,
+		}
+		wr.PartsOrder = []string{
+			zerolog.MessageFieldName,
+		}
+		lo := log.Output(wr)
+		lo.Info().Msg("Pass: " + e.Pass)
```

Comment:
If option --pass-filepath is specified, no need to output password into the log. Only print it when we are not creating file with the password.